--!strict

local testkit = require("./testkit")
local BENCH, START = testkit.benchmark()

local vide = require("../../vide")
local derive = vide.derive
local effect = vide.effect
local indexes = vide.indexes
local values = vide.values
local batch = vide.batch
local cleanup = vide.cleanup
local untrack = vide.untrack
local create = vide.create
local context = vide.context
local createSource = vide.createSource

assert(not vide.strict)

local function TITLE(name: string)
	print()
	print(testkit.color.white(name))
end

local function ROOT_BENCH(name: string, fn: () -> ())
	vide.root(function(destroy)
		BENCH(name, fn)
		return destroy
	end)()
end

local N = 2 ^ 20

TITLE("sources")

BENCH("create source", function()
	local cache = table.create(N)

	for i = 1, START(N) do
		cache[i] = createSource(1) -- stores getter (first return)
	end
end)

BENCH("get value", function()
	local get = createSource(1)

	for i = 1, START(N) do
		get()
	end
end)

BENCH("set value", function()
	local get, set = createSource(1)

	for i = 1, START(N) do
		set(i)
	end
end)

ROOT_BENCH("derive 1 source", function()
	local cache = table.create(N)
	local get = createSource(1)

	for i = 1, START(N) do
		cache[i] = derive(function()
			return get()
		end)
	end
end)

ROOT_BENCH("derive 4 sources", function()
	local cache = table.create(N)
	local src = vide.createSource(1)
	local src2 = vide.createSource(2)
	local src3 = vide.createSource(3)
	local src4 = vide.createSource(4)

	for i = 1, START(N) do
		cache[i] = derive(function()
			return src() + src2() + src3() + src4()
		end)
	end
end)

TITLE("graphs")

ROOT_BENCH("update 1->1 graph", function()
	local get, set = createSource(1)

	local _derived = derive(function()
		return get()
	end)

	for i = 1, START(N) do
		set(i)
	end
end)

ROOT_BENCH("update 1->1 graph with cleanup", function()
	local get, set = createSource(1)

	derive(function()
		cleanup(function() end)
		return get()
	end)

	for i = 1, START(N) do
		set(i)
	end
end)

ROOT_BENCH("update 1->1000 graph", function()
	local get, set = createSource(-1)

	for i = 1, 1000 do
		derive(function()
			return get()
		end)
	end

	set(0)

	for i = 1, START(10) do
		set(i)
	end
end)

ROOT_BENCH("update 1->1->1->1...1000 graph", function()
	local get, set = createSource(-1)

	local last = get
	for i = 1, 1000 do
		local l = last
		last = derive(function()
			return l()
		end)
	end

	set(0)

	for i = 1, START(10) do
		set(i)
	end
end)

-- todo: why does it hang at 1k? it didn't before
ROOT_BENCH("update 500->1 graph", function()
	local srcs: { { get: () -> any, set: (any) -> () } } = {}
	for i = 1, 500 do
		local get, set = createSource(0)
		srcs[i] = { get = get, set = set }
	end

	derive(function()
		for i = 1, 500 do
			srcs[i].get()
		end
		return false
	end)

	for i = 1, START(1) do
		for idx = 1, 500 do
			srcs[idx].set(i)
		end
	end
end)

ROOT_BENCH("update 1000->1 graph (batched)", function()
	local srcs: { { get: () -> any, set: (any) -> () } } = {}
	for i = 1, 1000 do
		local get, set = createSource(0)
		srcs[i] = { get = get, set = set }
	end

	derive(function()
		for i = 1, 1000 do
			srcs[i].get()
		end
		return false
	end)

	for i = 1, START(1) do
		batch(function()
			for idx = 1, 1000 do
				srcs[idx].set(i)
			end
		end)
	end
end)

-- todo: optimize this case
ROOT_BENCH("update 1000 1->1 common extern. graph", function()
	local ext, setExt = createSource(-1)

	local srcs: { { get: () -> any, set: (any) -> () } } = {}
	for i = 1, 1000 do
		local get, set = createSource(0)
		srcs[i] = { get = get, set = set }
		derive(function()
			return srcs[i].get() + ext()
		end)
	end

	setExt(0)

	for i = 1, START(10) do
		for idx = 1, 1000 do
			srcs[idx].set(i)
		end
	end
end)

TITLE("property apply")

ROOT_BENCH("apply 0 properties", function()
	local apply = require("../src/apply")
	local instance = create("Frame")({})

	for i = 1, START(N) do
		apply(instance, {})
	end
end)

ROOT_BENCH("apply 8 properties", function()
	local apply = require("../src/apply")
	local instance = create("Frame")({})

	for i = 1, START(N) do
		apply(instance, {
			Text = i,
			Text2 = i,
			Text3 = i,
			Text4 = i,
			Text5 = i,
			Text6 = i,
			Text7 = i,
			Text8 = i,
		})
	end
end)

ROOT_BENCH("bind property", function()
	local apply = require("../src/apply")

	local instance = create("Frame")({})
	local src = createSource(1)

	for i = 1, START(N) do
		apply(instance, {
			Text = src,
		})
	end

	return nil
end)

ROOT_BENCH("update binding", function()
	local apply = require("../src/apply")

	local instance = create("Frame")({})
	local get, set = createSource(1)

	apply(instance, {
		Text = get,
	})

	for i = 1, START(N) do
		set(i)
	end

	return nil
end)

TITLE("switch()")

ROOT_BENCH("switch()", function()
	local M = 2 ^ 8

	local map = {}
	for i = 1, M do
		map[i] = function()
			return i
		end
	end

	local input, setInput = createSource(0)
	vide.switch(input)(map)

	for i = 1, START(N) do
		setInput(bit32.band(i, M - 1) + 1) -- i % m + 1
	end
end)

TITLE("indexes()")

N /= 1024

ROOT_BENCH("indexes() all new", function()
	local data = {}

	for i = 1, N do
		data[i] = i
	end

	local src = createSource(data)

	START(N)

	local _list = indexes(src, function(v, i)
		return {}
	end)

	return nil
end)

ROOT_BENCH("indexes() no change", function()
	local data = {}

	for i = 1, N do
		data[i] = i
	end

	local src, setSrc = createSource(data)

	local _list = indexes(src, function(v, i)
		return {}
	end)

	START(N)

	setSrc(data)

	return nil
end)

ROOT_BENCH("indexes() all change", function()
	local data = {}

	for i = 1, N do
		data[i] = i
	end

	local src, setSrc = createSource(data)

	local _list = indexes(src, function(v, i)
		return {}
	end)

	--setSrc(src()) -- fill double buffer

	for i, v in data do
		data[i] = v + 1
	end

	START(N)

	setSrc(data)
end)

ROOT_BENCH("indexes() all remove", function()
	local data = {}

	for i = 1, N do
		data[i] = i
	end

	local src, setSrc = createSource(data)

	local _list = indexes(src, function(v, i)
		return {}
	end)

	table.clear(data)

	START(N)

	setSrc(data)

	return nil
end)

TITLE("values()")

ROOT_BENCH("values() all new", function()
	local data = {}

	for i = 1, N do
		data[i] = {}
	end

	local src = createSource(data)

	START(N)

	local _list = values(src, function(v, i)
		return {}
	end)

	return nil
end)

ROOT_BENCH("values() no change", function()
	local data = {}

	for i = 1, N do
		data[i] = {}
	end

	local src, setSrc = createSource(data)

	local _list = values(src, function(v, i)
		return {}
	end)

	setSrc(src()) -- fill double buffer

	START(N)

	setSrc(data)
end)

ROOT_BENCH("values() all change", function()
	local data = {}

	for i = 1, N do
		data[i] = {}
	end

	local src, setSrc = createSource(data)

	local _list = values(src, function(v, i)
		return {}
	end)

	setSrc(src()) -- fill double buffer

	for i = 1, N do
		local r = math.random(1, #data)
		data[i], data[r] = data[r], data[i]
	end

	START(N)

	setSrc(data)
end)

ROOT_BENCH("values() all remove", function()
	local data = {}

	for i = 1, N do
		data[i] = {}
	end

	local src, setSrc = createSource(data)

	local _list = values(src, function(v, i)
		return {}
	end)

	table.clear(data)

	START(N)

	setSrc(data)
end)

TITLE("context()")

ROOT_BENCH("set context", function()
	local ctx = context()

	for i = 1, START(N) do
		ctx(i, function() end)
	end
end)

ROOT_BENCH("get context (depth=1)", function()
	local ctx = context()

	local function run()
		for i = 1, START(N) do
			ctx()
		end
	end

	ctx(1, function()
		run()
	end)
end)

local depth = 10
ROOT_BENCH(`get context (depth={depth})`, function()
	local ctx = context()

	local function run()
		for i = 1, START(N) do
			ctx()
		end
	end

	local function nest_effect(fn)
		untrack(function()
			effect(fn)
			return nil
		end)
	end

	local f = run
	for i = 1, depth - 1 do
		local f_inner = f
		f = function()
			nest_effect(f_inner)
		end
	end

	ctx(1, function()
		f()
	end)
end)

TITLE("spring()")

ROOT_BENCH("spring update", function()
	local root, createSource2, spring = vide.root, vide.createSource, vide.spring

	local src, setSrc = createSource2(0)

	root(function()
		for i = 1, N do
			spring(src)
		end

		START(N)

		setSrc(1)

		return nil
	end)
end)

ROOT_BENCH("spring step", function()
	local root, createSource2, spring = vide.root, vide.createSource, vide.spring

	local src, setSrc = createSource2(0)

	root(function()
		for i = 1, N do
			spring(src)
		end

		setSrc(1)

		START(N)

		vide.step(1 / 60)

		return nil
	end)
end)

return nil
